mithril-client/src/cardano_database_client/proving.rs

```diff
@@ -0,0 +1,617 @@
+                match ImmutableFile::new(Path::new(immutable_file_name).to_path_buf()) {
+                    Ok(immutable_file) => immutable_file.number <= last_immutable_file_number,
+                    Err(_) => false,
+                }
+            })
+            .map(|(_immutable_file_name, digest)| digest)
+            .collect::<Vec<_>>();
+        let merkle_tree: MKTree<MKTreeStoreInMemory> = MKTree::new(&downloaded_digests_values)?;
+        let immutable_digester = CardanoImmutableDigester::new(network, None, self.logger.clone());
+        let computed_digests = immutable_digester
+            .compute_digests_for_range(database_dir, &immutable_file_number_range)
+            .await?
+            .entries
+            .values()
+            .map(MKTreeNode::from)
+            .collect::<Vec<_>>();
+        delete_directory(&Self::digest_target_dir(database_dir))?;
+
+        merkle_tree.compute_proof(&computed_digests)
+    }
+
+    async fn download_unpack_digest_file(
+        &self,
+        locations: &[DigestLocation],
+        digest_file_target_dir: &Path,
+    ) -> MithrilResult<()> {
+        create_directory_if_not_exists(digest_file_target_dir)?;
+        let mut locations_sorted = locations.to_owned();
+        locations_sorted.sort();
+        for location in locations_sorted {
+            let download_id = MithrilEvent::new_cardano_database_download_id();
+            self.feedback_sender
+                .send_event(MithrilEvent::CardanoDatabase(
+                    MithrilEventCardanoDatabase::DigestDownloadStarted {
+                        download_id: download_id.clone(),
+                    },
+                ))
+                .await;
+            let file_downloader = match &location {
+                DigestLocation::CloudStorage { .. } | DigestLocation::Aggregator { .. } => {
+                    self.http_file_downloader.clone()
+                }
+            };
+            let file_downloader_uri: FileDownloaderUri = location.into();
+            let downloaded = file_downloader
+                .download_unpack(
+                    &file_downloader_uri,
+                    digest_file_target_dir,
+                    None,
+                    DownloadEvent::Digest {
+                        download_id: download_id.clone(),
+                    },
+                )
+                .await;
+            match downloaded {
+                Ok(_) => {
+                    self.feedback_sender
+                        .send_event(MithrilEvent::CardanoDatabase(
+                            MithrilEventCardanoDatabase::DigestDownloadCompleted { download_id },
+                        ))
+                        .await;
+                    return Ok(());
+                }
+                Err(e) => {
+                    slog::error!(
+                        self.logger,
+                        "Failed downloading and unpacking digest for location {file_downloader_uri:?}"; "error" => e.to_string()
```

Comment:
Should we debug the error to show the stack trace ? Else only the main message will be shown.
```suggestion
                        "Failed downloading and unpacking digest for location {file_downloader_uri:?}"; "error" => ?e
```